File: backend/routes/devices.js
/**
 * ClockMQTT — Device Routes
 * ==========================
 * /api/devices — CRUD for device registration and status.
 */

const { Router } = require('express');
const deviceManager = require('../deviceManager');

const router = Router();

// GET /api/devices — list all devices with status
router.get('/', (req, res) => {
  const devices = deviceManager.getAllDevices();
  res.json({ devices, total: devices.length });
});

// POST /api/devices — register a new device
router.post('/', (req, res) => {
  try {
    const { device_id, device_name, device_key } = req.body;

    if (!device_id) {
      return res.status(400).json({ detail: 'device_id is required' });
    }
    if (!device_key || device_key.length < 8) {
      return res.status(400).json({ detail: 'device_key must be at least 8 chars' });
    }

    deviceManager.registerDevice({ device_id, device_name, device_key });

    res.json({
      status: 'ok',
      device_id,
      device_name: device_name || device_id,
      message: 'Device registered. Add credentials to broker/server.js USERS object.',
    });
  } catch (e) {
    res.status(400).json({ detail: e.message });
  }
});

// GET /api/devices/:id — get single device detail
router.get('/:id', (req, res) => {
  const device = deviceManager.getDevice(req.params.id);
  if (!device) {
    // Return default for unknown devices (matches Python behavior)
    return res.json({
      device_id: req.params.id,
      device_name: req.params.id,
      online: false,
      rssi: null,
      fw_ver: null,
    });
  }
  res.json(device);
});

// DELETE /api/devices/:id — remove a device
router.delete('/:id', (req, res) => {
  deviceManager.removeDevice(req.params.id);
  res.json({ status: 'ok', message: `Device ${req.params.id} removed` });
});

module.exports = router;

File: backend/routes/devices.py
"""Device Management API."""

from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from sqlalchemy import text
from database import get_db
from models.schemas import DeviceCreate, DeviceResponse
import json

router = APIRouter(prefix="/api/devices", tags=["devices"])

# In-memory device status (updated by MQTT status messages)
_device_status: dict[str, dict] = {}


# Called by MQTT client when device status changes
def update_device_status(topic: str, payload: str) -> None:
    # Topic: inkpad/{device_id}/status
    parts = topic.split("/")
    if len(parts) >= 3:
        device_id = parts[1]
        try:
            data = json.loads(payload)
            _device_status[device_id] = {
                "online": data.get("online", False),
                "rssi": data.get("rssi"),
                "fw_ver": data.get("fw_ver"),
                "last_seen": None,  # would use datetime.now()
            }
        except json.JSONDecodeError:
            pass


@router.get("")
def list_devices():
    """List all registered devices with status."""
    # In production: read from DB, merge with _device_status
    devices = [
        {
            "device_id": "Clock1",
            "device_name": "Main Clock",
            "online": _device_status.get("Clock1", {}).get("online", False),
            "rssi": _device_status.get("Clock1", {}).get("rssi"),
            "fw_ver": _device_status.get("Clock1", {}).get("fw_ver"),
        }
    ]
    return {"devices": devices, "total": len(devices)}


@router.post("")
def add_device(req: DeviceCreate):
    """Register a new device."""
    # In production: save to DB, generate device_key
    return {
        "status": "ok",
        "device_id": req.device_id,
        "device_name": req.device_name,
        "message": "Device registered. Add credentials to broker/server.js USERS object.",
    }


@router.get("/{device_id}")
def get_device(device_id: str):
    """Get device detail."""
    status = _device_status.get(device_id, {})
    return {
        "device_id": device_id,
        "device_name": "Clock1" if device_id == "Clock1" else device_id,
        "online": status.get("online", False),
        "rssi": status.get("rssi"),
        "fw_ver": status.get("fw_ver"),
    }


@router.delete("/{device_id}")
def delete_device(device_id: str):
    """Delete a device."""
    _device_status.pop(device_id, None)
    return {"status": "ok", "message": f"Device {device_id} removed"}
